pkg/clients/elasticache/elasticache.go

```diff
@@ -89,209 +51,267 @@ func NewReplicationGroupDescription(g *v1alpha2.ReplicationGroup) string {
+		EngineVersion:              g.EngineVersion,
+		NotificationTopicArn:       g.NotificationTopicARN,
+		NumCacheClusters:           aws.Int64Address(g.NumCacheClusters),
+		NumNodeGroups:              aws.Int64Address(g.NumNodeGroups),
+		Port:                       aws.Int64Address(g.Port),
+		PreferredCacheClusterAZs:   g.PreferredCacheClusterAZs,
+		PreferredMaintenanceWindow: g.PreferredMaintenanceWindow,
+		PrimaryClusterId:           g.PrimaryClusterID,
+		ReplicasPerNodeGroup:       aws.Int64Address(g.ReplicasPerNodeGroup),
+		SecurityGroupIds:           g.SecurityGroupIDs,
+		SnapshotArns:               g.SnapshotARNs,
+		SnapshotName:               g.SnapshotName,
+		SnapshotRetentionLimit:     aws.Int64Address(g.SnapshotRetentionLimit),
+		SnapshotWindow:             g.SnapshotWindow,
+		TransitEncryptionEnabled:   g.TransitEncryptionEnabled,
 	}
-}
-
-func newNodeGroupConfigurations(g *v1alpha2.ReplicationGroup) []elasticache.NodeGroupConfiguration {
-	if len(g.Spec.NodeGroupConfiguration) == 0 {
-		return nil
+	if len(g.Tags) != 0 {
+		c.Tags = make([]elasticache.Tag, len(g.Tags))
+		for i, tag := range g.Tags {
+			c.Tags[i] = elasticache.Tag{
+				Key:   &tag.Key,
+				Value: &tag.Value,
+			}
+		}
 	}
-	nc := make([]elasticache.NodeGroupConfiguration, len(g.Spec.NodeGroupConfiguration))
-	for i, cfg := range g.Spec.NodeGroupConfiguration {
-		nc[i] = elasticache.NodeGroupConfiguration{
-			PrimaryAvailabilityZone:  aws.String(cfg.PrimaryAvailabilityZone),
-			ReplicaAvailabilityZones: cfg.ReplicaAvailabilityZones,
-			ReplicaCount:             aws.Int64(cfg.ReplicaCount),
-			Slots:                    aws.String(cfg.Slots),
+	if len(g.NodeGroupConfiguration) != 0 {
+		c.NodeGroupConfiguration = make([]elasticache.NodeGroupConfiguration, len(c.NodeGroupConfiguration))
+		for i, cfg := range g.NodeGroupConfiguration {
+			c.NodeGroupConfiguration[i] = elasticache.NodeGroupConfiguration{
+				PrimaryAvailabilityZone:  cfg.PrimaryAvailabilityZone,
+				ReplicaAvailabilityZones: cfg.ReplicaAvailabilityZones,
+				ReplicaCount:             aws.Int64Address(cfg.ReplicaCount),
+				Slots:                    cfg.Slots,
+			}
 		}
 	}
-	return nc
+	return c
 }
 
 // NewModifyReplicationGroupInput returns ElastiCache replication group
 // modification input suitable for use with the AWS API.
-func NewModifyReplicationGroupInput(g *v1alpha2.ReplicationGroup) *elasticache.ModifyReplicationGroupInput {
+func NewModifyReplicationGroupInput(g v1alpha2.ReplicationGroupParameters, id string) *elasticache.ModifyReplicationGroupInput {
 	return &elasticache.ModifyReplicationGroupInput{
-		ReplicationGroupId: aws.String(NewReplicationGroupID(g), aws.FieldRequired),
-
-		// TODO(negz): Should this be a configurable part of the replication
-		// group spec? If we did wait until the next maintenance window to apply
-		// changes we'd need some way to account for the pending changes during
-		// our sync logic.
```

Comment:
I think that removing this TODO makes sense given that deferred apply doesn't really make sense for Crossplane resources, but I wonder if we should document this behaviour somehow.